enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,216 @@
+operator authors and pipeline developers can reason about the structure of an index in a standard way that's
+also consumable by continuous integration tools. 
+
+## Motivation
+
+Currently, a lot of the rules about the structure of an index are implicitly documented in various sources that are scattered. This makes 
+it hard for operator authors and pipeline developers creating indexes to rapidly make decisions that impact the graph of the index. To alleviate 
+some of the pain points related to that, a number of tools have cropped up (operator-courier, opm etc) to provide users a way to validate their bundles
+and indexes. However, like all tools that aspire to provide data validation, these tools, or api libraries that these tools import rules from, have to 
+be constantly maintained to keep up with the latest rule changes, while also maintaining older rules in previous releases. This introduces a lot of overhead
+in ensuring that olm is easy to use by users, while also leaving windows for lapses due to mis-coordination between different tools across different versions. 
+
+CUE is a configuration and data validation language that makes it easy to define the structure and validate the data with the tools it provides.  
+
+### Goals
+
+1. Configure CUE definition for an index
+
+### Non-Goals
+
+
+## Proposal
+
+### Implementation Details
+
+An index will be represented as a yaml/json: 
+
+For example: 
+
+```index.yaml
+graph:
+  nodes:
+    - bundlePath: quay.io/etcd.v1.0.0
+      provides:
+        - type: olm.package
+          version: 1.0.0
+          packageName: etcd
+        - type: olm.gvk
+          version: v1beta2
+          group: etcd.database.coreos.com
+          kind: EtcdCluster
+        - type: label
+          value: alpha
+      constraints:
+        - - type: label
+            value: LTS
+    - bundlePath: quay.io/etcd.v1.0.1
+      provides:
+        - type: olm.package
+          version: 1.0.1
+          packageName: etcd
+        - type: olm.gvk
+          version: v1beta2
+          group:etcd.database.coreos.com
+          kind: EtcdCluster
+      constraints:
+        - - type: label
+            value: LTS
+    - bundlePath: quay.io/vault.1.0.1
+      provides:
+        - type: olm.package
+          version: 1.0.1
+          packageName: vault
+        - type: olm.gvk
+          version: v1beta2
+          group: vault.database.coreos.com
+          kind: VaultCluster
+        - type: label
+          value: stable
+      constraints:
+        - - type: olm.gvk
+            version: v1beta2
+            group: etcd.database.coreos.com
+            kind: etcd
+  schema: 1.0
+  edges:
+    - quay.io/etcd.v1.0.1:
+        - quay.io/etcd.v1.0.0
+```
+
+A new opm sub-command, `represent` will be introduced that will output a yaml or a json representation of an index. 
+
+I.e, `opm represent --output=yaml docker.io/my-namespace/my-index` will output the `index.yaml` in the example above.
```

Comment:
Who is the audience for this command? Who consumes the output? What is the driving use case behind it?

Reply:
> Who is the audience for this command? 
Index authors/pipeline 

> Who consumes the output? 
See [goals](https://github.com/operator-framework/enhancements/blob/80852d59b1b03d6328e1ce03f353a23a035aca29/enhancements/opm-content-management.md#goals) mentioned in the OPM content management enhancement, which is the ultimate goal. 
Specifically, `opm users can validate their bundles in an environment without docker deamon` and `opm users can inspect the catalog content and return the update graphs of Operators in machine readable formats` 

> What is the driving use case behind it?

Again the goals of the opm content management enhancement has the answers. It's making me think that maybe it's a good idea to link the content management enchantment in this enhancement to make it clear that this one's just the building block for the actual use cases mentioned in that enhancement. 

(PS: I did have a thought of moving some of the content over from the content management enhancement to this one, but on second thought having them decoupled is probably preferable)

Reply:
Gotcha. Maybe it's worth merging the two back together. At least the link should be obvious. Without it, there is a lot of context loss when looking at this enhancement.